File: python/leetcode/004.py
class Solution(object):
    def findMedianSortedArrays(self, nums1, nums2):
        m, n = len(nums1), len(nums2)
        if (m + n) % 2 != 0:
            return self.getKth(nums1, nums2, (m + n) / 2 + 1)
        else:
            return (self.getKth(nums1, nums2, (m + n) / 2) + self.getKth(nums1, nums2, (m + n) / 2 + 1)) / 2.0

    def getKth(self, nums1, nums2, k):
        m, n = len(nums1), len(nums2)
        if m > n:
            return self.getKth(nums2, nums1, k)
        if m == 0:
            return nums2[k - 1]
        if k == 1:
            return min(nums1[0], nums2[0])
        idxOfNum1 = min(m, k / 2)
        idxOfNum2 = k - idxOfNum1
        if nums1[idxOfNum1 - 1] < nums2[idxOfNum2 - 1]:
            return self.getKth(nums1[idxOfNum1:], nums2, idxOfNum2)
        else:
            return self.getKth(nums1, nums2[idxOfNum2:],  idxOfNum1)

File: cpp/leetcode/004.cpp
class Solution {
public:
    int getKth(vector<int> &nums1, vector<int> &nums2, int k) {
        int m = nums1.size();
        int n = nums2.size();
        if (m > n) {
            return getKth(nums2, nums1, k);
        }
        if (m == 0) {
            return nums2[k - 1];
        }
        if (k == 1) {
            return min(nums1[0], nums2[0]);
        }
        int pa = min(k / 2, m), pb = k - pa;
        if (nums1[pa - 1] < nums2[pb - 1]) {
            vector<int> tmp(nums1.begin() + pa, nums1.end());
            return getKth(tmp, nums2, k - pa);
        } else {
            vector<int> tmp(nums2.begin() + pb, nums2.end());
            return getKth(nums1, tmp, k - pb);
        }
    }

    double findMedigetKthortedArrays(vector<int> &nums1, vector<int> &nums2) {
        int m = nums1.size(), n = nums2.size();
        int k = (n + m) / 2;
        if ((m + n) % 2 != 0) {
            return getKth(nums1, nums2, k + 1);
        } else {
            return (getKth(nums1, nums2, k) + getKth(nums1, nums2, k + 1)) / 2.0;
        }
    }
};